Implement Installation Process Domain Rules

redi-capoj opened the pull request:
Implements the domain model for managing device installation workflows, handling the complete lifecycle from work order creation to installation completion. Includes test coverage for business rules including boot notification handling, device assignment flows, and process completion validation. Implements key rules  around finished process handling, and boot notification confirmation requirements.

redi-capoj committed:
test(InstallationProcess): create InstallationProcess test

- Create InstallationProcessTest class-
- Set up the environment by creating a new process with a given order id.
- Add test for preventing duplicate installation processes.
- Verify test failure.

redi-capoj committed:
feat(InstallationProcess): create InstallationProcess class

- Create InstallationProcess class.
- Add constructor with argument passed: orderId.
- Handle the creation of only one instance of InstallationProcess with a given orderId.
- Should throw exception if the passed orderId already exists.
- Verify that test succeeds.

redi-capoj committed:
test(InstallationProcess): add tests when process is finished.

- Add test method to prevent assigning of device after process is finished.
- Add test method to prevent assigning of installer after process is finished.

redi-capoj committed:
feat(InstallationProcess): implement assigner methods.

- Implement methods in InstallerProcess.
- Add necessary logic to prevent assigning when process is finished.
- Verify that tests pass successfully.

redi-capoj committed:
test(InstallationProcess): add tests for BootNotification.

- Add test for matching boot notification device with assigned device.
- Add test for ignoring boot notification when process is finished.

redi-capoj committed:
feat(InstallationProcess): implement receiveBootNotification.

- Add and implement receiveBootNotification.
- Create BootNotification record.
- Verify that tests pass successfully.

redi-capoj committed:
test(InstallationProcess): add test for updating boot notification or assigning new device..

- Add test method to require new confirmation if boot notification is updated.
- Add test method to require new confirmation when reassigning device in the process.

redi-capoj committed:
feat(InstallationProcess): add methods for handling boot notification update.

- Implement confirming boot notification method.
- Add update boot notification method and handling necessary logic.
- Add getter method for accessing the status of boot notification.
- Verify that tests pass as expected.

redi-capoj committed:
refactor(test): Refactor InstallationProcessTest

- Add test case for creating a new process.
- Verify test failure.

redi-capoj committed:
refactor(InstallationProcess): Refactor InstallationProcess

- Rework the implementation.

redi-capoj committed:
feature(process): add WorkOrder and Ownership immutable classes

redi-capoj committed:
refactor(test): refactor test cases when process is finished

- Verify tests failure.

redi-capoj committed:
refactor(process): refactor implementation of InstallationProcess

- Add helper method to validate if process is finished or throw exception.
- Implement assigner method by verifying process is not finished first.
- Implemented finish method to finish the state of the process.
- Verify tests passing.

redi-capoj committed:
test(InstallationProcess): add test cases when receiving boot notification

- Refactor test method to ignore boot notification when process finished.
- Add test case when receiving notification without confirming, should be stored as pending.

redi-capoj committed:
refactor(InstallationProcess): refactor receiving boot

- Add pendingBootNotification field to handle logic when boot is not yet confirmed.
- Refactor receiveBootNotification to ignore notification when process is finished or does not  equal the device id.
- Verified tests pass successfully.

redi-capoj committed:
test(InstallationProcess): add boot notification test cases

- Refactor test case to throw exception when device id and notifications device id does not match or device is not assigned.
- Add test cases to not confirm boot without receiving it first.
- Add test case to clear pending notification and to store it as confirmed notification if boot is confirmed correctly.

redi-capoj committed:
feat(InstallationProcess): confirm boot logic

- Implement confirm boot method to confirm the boot notification if correct.
- Confirmed notification should be set to the pending notification value and pending notification should be reset to null.
_ Refactor receive notification to throw exception if device id not match or is not assigned.

redi-capoj committed:
test(notification): add test cases for boot notification

- Add test case to verify boot is not confirmed if we update any boot notification data.
- Add test to check that no new confirmation is needed if the second notification is exactly the same as the previous.
- Add test case to check when reassigning a new device should require a new boot notification and confirmation.

redi-capoj committed:
refactor(installation): add logic when assigning new device

- Implement helper method resetBootNotification to reset boot notification and require a new confirmation.
- Add method resetBootNotification after reassigning device, to require new notification & confirmation.
- Add getter method for device id.
- Fix typo in test method.

redi-capoj committed:
test(location): add test cases for location

- Add location object with defined data.
- Add test case, location cannot be set if boot notification is not confirmed.
- Add test case to verify that after assigning new device, the boot notification, confirmation and location are reset.

redi-capoj committed:
feat(location): create Location

- Create Location record (immutable class)
- Use Lombok @Builder.

redi-capoj committed:
feat(location): implement setLocation

- Add setLocation method to correctly set location if process is not finished.
_ Check if boot is confirmed, if not throw exception.
- Check if the current location matches with the new one to not override it.
- Rename resetBootNotification method to resetBootState, in which the location is reset too if device is reassigned.

redi-capoj committed:
test(finish-installation): add test cases for finishing installation

- Add test case to verify that exception is thrown if process finishes when misses data.
- Add test case to verify that process finishes correctly if there are not missing data.

redi-capoj committed:
feat(finish-installation): refactor finish method

- Add in finish method helper method validateCompletedInstallation to validate the completion of process, if any fields are missing an exception is thrown.

redi-capoj committed:
chore(BootNotification): remove redundant constructor

redi-capoj committed:
feat(Fixture): add InstallationProcessFixture helper class

- Create helper class to simplify tests.
- Add static methods which returns data for tests.
- Add processInProgress method which contains a complete installation process.

redi-capoj committed:
refactor(project): refactor given data

- Refactor needed data to simplify the code, and to use predefined methods with data.
- Fix typos and code formatting.

redi-capoj committed:
refactor(InstallationProcess): add @Getter annotation

- Replace getter methods with lombok annotation @Getter for more clean code.
- Format code.
- Fix any typos.

redi-capoj committed:
style(InstallationProcess): reformat code and add comments

github-actions[bot] committed:
Autogenerated JaCoCo coverage badge

michal-michaluk commented on src/main/java/devices/configuration/installation/InstallationProcess.java:
```diff
@@ -0,0 +1,103 @@
+    private void resetBootState() {
+        confirmedBootNotification = null;
+        pendingBootNotification = null;
+        location = null;
```
That could be clarified with business experts. But I respect taking that into consideration, and it make total sense. 👍

michal-michaluk commented on src/main/java/devices/configuration/installation/InstallationProcess.java:
```diff
@@ -0,0 +1,103 @@
+
+    public void setLocation(Location someLocation) {
+        validateNotFinished();
+        if (confirmedBootNotification == null) {
```
Interesting rule, not sure it is necessary to force conforming boot before setting location. But for sure it can be forced like that.

I see the order on Event Storming, but could be clarified with bizz experts.

Great job 👏 (no changes needed)

michal-michaluk replied:
ok lets decide it is rule business needs

michal-michaluk commented on src/test/java/installation/InstallationProcessFixture.java:
```diff
@@ -0,0 +1,63 @@
+    private static final String deviceId = "ALF-83266831";
+    private static final String installerId = "INSTALLER-456";
+
+    public static BootNotification defaultNotification() {
```
Default is not proper word, it is example or some, any etc.

We can name the example by some key property ex. boot for 2.0 protocol or something like that.
Especially clear when we need second example - then name to express the difference

michal-michaluk commented on src/test/java/installation/InstallationProcessFixture.java:
```diff
@@ -0,0 +1,63 @@
+
+    private static final String orderId = "K56F";
+    private static final String deviceId = "ALF-83266831";
+    private static final String installerId = "INSTALLER-456";
```
No way to change that can be a bit limiting, random will not work as wall.

It is possible to do state full Fixtures: constructor / factory method / builder to setup such values, then in each method we don't need to specify Ids or other repeated values.

It will be refactored later when we will test with two work orders or installation processes

michal-michaluk commented on src/test/java/installation/InstallationProcessTest.java:
```diff
@@ -0,0 +1,234 @@
+        someLocation = someLocation();
+        workOrder = defaultOrder();
+
+        process = new InstallationProcess(workOrder);
```
That value is overridden in almost all tests. Better to keep it local variable for each test

michal-michaluk commented on src/test/java/installation/InstallationProcessTest.java:
```diff
@@ -0,0 +1,234 @@
+        // then exception should be thrown if we send notification
+        process.assignDevice("wrong-device");
+
+        assertThrows(IllegalStateException.class, () -> process.receiveBootNotification(defaultBoot));
```
I think it deserves to be separate test.

michal-michaluk commented on src/test/java/installation/InstallationProcessTest.java:
```diff
@@ -0,0 +1,234 @@
+        // given
+        process.assignDevice(deviceId);
+        process.receiveBootNotification(defaultBoot);
+        process.confirmBoot();
```
That is // when section

michal-michaluk commented on src/test/java/installation/InstallationProcessTest.java:
```diff
@@ -0,0 +1,234 @@
+        // when receiving boot notification
+        process.receiveBootNotification(defaultBoot);
+
+        // then setting location if boot notification is not confirmed
```
Not sure it is requirement, but for sure it matches process flow discovered on Event Storming

michal-michaluk commented on src/test/java/installation/InstallationProcessTest.java:
```diff
@@ -0,0 +1,234 @@
+        // then boot notification, confirmation & location are needed again
+        assertNull(process.getConfirmedBootNotification());
+        assertNull(process.getPendingBootNotification());
+        assertNull(process.getLocation());
```
That part NEED to be clarified with bizz, do we need reset location if we started with wrong device ...

Interesting scenario to discuss

michal-michaluk replied:
ok so lets decide that bigness needs that domain rule

michal-michaluk commented on src/test/java/installation/InstallationProcessTest.java:
```diff
@@ -0,0 +1,234 @@
+
+    @Test
+    void whenFinishingWithoutCompleteTerms_thenThrowException() {
+        // cannot finish process when missing:
```
Somehow having that installation process already initialized in field, leaves empty given and and make my brain always asking what's the starting state ...

Having given with creating fresh or almost finished process will be more natural.

michal-michaluk commented on src/test/java/installation/InstallationProcessTest.java:
```diff
@@ -0,0 +1,234 @@
+    @Test
+    void whenFinishingWithAllCompletedTerms_thenMarkAsFinished() {
+        // given a completed process
+        process = processInProgress();
```
Comment can suggest better name for `processInProcess` method
`almostCompletedProcess` ...
But I enjoy current name, kind of funny

michal-michaluk commented on src/test/java/installation/InstallationProcessFixture.java:
```diff
@@ -0,0 +1,63 @@
+                .build();
+    }
+
+    public static InstallationProcess processInProgress() {
```
Please find more better name

michal-michaluk commented on src/main/java/devices/configuration/installation/InstallationProcess.java:
```diff
@@ -0,0 +1,103 @@
+
+    private boolean isFinished;
+
+    public InstallationProcess(WorkOrder workOrder) {
```
For persistence we will need AllArgsConstructor.

I like named constructors (static factory methods) for constructions driven by domain in opposite to persistence / adapter driven creations.

But current state does a job for current scope (no adapters) of solution.

michal-michaluk commented on src/main/java/devices/configuration/installation/InstallationProcess.java:
```diff
@@ -0,0 +1,103 @@
+import java.util.Objects;
+
+@Getter
+public class InstallationProcess {
```
Please add DomainEvent's for each step. Having history of such process in DB is always a nice feature.
And input for process improvements.

michal-michaluk commented:
@redicapoj In general it looks VERY good.
I left some comments just want to say “it is nice 👍”, and no change is required.

Some rules should be discussed with biznes, but it is normal that we discover some rules and need to confirm them.
Independently to final decision, work flow and code quality is great.

Most doubts I have in test data handling:
- constants in Fixture
- ⁠instance of proces in test field
- ⁠naming of fixture methods

And extension of InstallationProcess to emit events will be useful in future.

redi-capoj committed:
refactor(fixture): refactor code

- Remove the constants to give more flexibility in tests.
- Inspired from PR #9 solved by Michal and Gisela:
   -> Replace the "processInProgress" method with "given" method to create a custom process with a given id, we use Consumer functional interface.
 - Rename from defaultNotification to a more specific name for this boot notification: "ultraChargeBootNotification".

redi-capoj committed:
refactor(test): refactor test class

- Remove process from setUp to not override in each test case, use as local variable.
- Replace construction of Installation Process with the use of static factory method "create" which create a new process.
- Make use of helper method "given" to construct a new process with some data.
- Separate test cases for not assigning and not matching device.
- Improve comments.

redi-capoj committed:
refactor(InstallationProcess): replace constructor

- Add lombok @AllArgsConstructor.
- Replace constructor for creation of new process with static method "create" with a given order.
- Verify all tests pass successfully.

redi-capoj committed:
test(events): add assertions for events

- Add assertions to confirm events are emitted and stored correctly.
- Add case to check no events are emitted if the data is the same.
- Clear events stored during the construction of given method, to only emit events afterward.

redi-capoj committed:
add(DomainEvent): create DomainEvent interface

- Add domain event interface.
- Add specified events to store each event with the work order and its specific data.

redi-capoj committed:
feat(process): add event storing functionality

- Add list to store each event for the created process.
- Add logic to emit event in each called method.

redi-capoj committed:
refactor(reset-boot): refactor emitting events after reset

- Add conditions to check before resetting the boot notification, confirmation and location.
- Events should be emitted only if the value was not null before.

github-actions[bot] committed:
Autogenerated JaCoCo coverage badge

michal-michaluk commented on src/main/java/devices/configuration/installation/DomainEvent.java:
```diff
@@ -0,0 +1,19 @@
+public interface DomainEvent {
+    record ProcessCreated(WorkOrder workOrder) implements DomainEvent {}
+
+    record InstallerAssigned(WorkOrder workOrder, String installerId) implements DomainEvent {}
+
+    record DeviceAssigned(WorkOrder workOrder, String deviceId) implements DomainEvent {}
+
+    record BootNotificationReceived(WorkOrder workOrder, BootNotification pendingBootNotification) implements DomainEvent {}
+
+    record BootNotificationConfirmed(WorkOrder workOrder, BootNotification confirmedBootNotification)implements DomainEvent {}
+
+    record LocationChanged(WorkOrder workOrder, Location location) implements DomainEvent {}
+
+    record InstallationFinished(WorkOrder workOrder) implements DomainEvent {}
```
WorkOrder.orderId will be enough, we don't need to repeat entire WorkOrder for each event, just id + changed data.

ProcessCreated is kind of unique, here WorkOrder is a "changed" data

michal-michaluk commented on src/main/java/devices/configuration/installation/InstallationProcess.java:
```diff
@@ -0,0 +1,134 @@
+        }
+    }
+
+    public WorkOrder workOrder() {
```
that can be removed after domain events are simplified

michal-michaluk commented on src/test/java/installation/InstallationProcessTest.java:
```diff
@@ -0,0 +1,414 @@
+    @Test
+    void whenAssignInstallerAfterFinish_thenThrowException() {
+        // given - a completed process
+        var process = given(workOrder, p -> {
```
clever, you looked into PR of Gisela, very nice 👏🏻

michal-michaluk commented on src/main/java/devices/configuration/installation/InstallationProcess.java:
```diff
@@ -0,0 +1,134 @@
+import java.util.List;
+import java.util.Objects;
+
+@Getter
```
It will be good to find a way to  test that object without getters.

Lets imagine some Value Object useful on Frontend showing current state of installation.
Gives us opportunity to present:
✅ responsible installer assigned
✅ device assigned
✅ BootNotification recieved
⬜️ BootNotification confirmed
...

and allows to view / review details of WorkOrder, BootNotification etc.

that ValueObject can be then produced by Process and verified in tests

michal-michaluk reviewed (changes requested):
Great progress, and some next improvements proposed

redi-capoj committed:
refactor(events): replace the workOrder argument with order id

- Simplify emitting events to store only order id and changed data.

redi-capoj committed:
refactor(events): replace the workOrder argument with order id

- Simplify emitting events to store only order id and changed data.

redi-capoj committed:
refactor(events): replace the workOrder argument with order id

- Simplify emitting events to store only order id and changed data.
- Remove helper method "workOrder()"

redi-capoj committed:
refactor(process): test current state of installation process

- Test the state of the installation process by capturing the current state of it in an immutable object of type "InstallationState"
- Refactor the assertions which checked the process with the new approach.
- Move the installation test package to the correct destination.

redi-capoj committed:
refactor(process): implement getter for current state of installation.

- Remove @Getter annotation which was used before.
- Implement new method "getCurrentState()" which construct a new InstallationState with the current values of the process.

redi-capoj committed:
feat(installation-state): create record InstallationState

- Create immutable class InstallationState.

github-actions[bot] committed:
Autogenerated JaCoCo coverage badge

redi-capoj commented:
> Great progress, and some next improvements proposed

Thank you, implemented the proposed improvements.
